code/datums/components/attachment_handler.dm

```diff
@@ -0,0 +1,304 @@
+	. = ..()
+	if(!isitem(parent))
+		return COMPONENT_INCOMPATIBLE
+
+	slots = _slots
+	attachables_allowed = _attachables_allowed
+	on_attach = _on_attach
+	on_detach = _on_detach
+	attachment_offsets = _attachment_offsets
+	attachable_overlays = overlays //This is incase the parent wishes to have a stored reference to this list.
+	attachable_overlays += slots 
+	overlay_icon_states += slots 
+
+	if(length(starting_attachmments)) //Attaches starting attachments
+		for(var/starting_attachment_type in starting_attachmments)
+			finish_handle_attachment(new starting_attachment_type())
+
+	update_parent_overlay()
+
+	RegisterSignal(parent, COMSIG_PARENT_ATTACKBY , .proc/start_handle_attachment)
+	RegisterSignal(parent, COMSIG_ATOM_UPDATE_OVERLAYS, .proc/update_parent_overlay)
+
+	RegisterSignal(parent, COMSIG_CLICK_ALT, .proc/start_detach)
+	RegisterSignal(parent, COMSIG_ITEM_ACTIVATE_ATTACHMENT, .proc/activate_attachment)
+	RegisterSignal(parent, COMSIG_ITEM_ATTACH_WITHOUT_USER, .proc/attach_without_user)
+	RegisterSignal(parent, COMSIG_ITEM_UPDATE_ATTACHMENT_ICON, .proc/overlay_icon_update)
+	RegisterSignal(parent, COMSIG_PARENT_QDELETING, .proc/clean_references)
+
+///Starts processing the attack, and whether or not the attachable can attack.
+/datum/component/attachment_handler/proc/start_handle_attachment(datum/source, obj/attacking, mob/attacker)
+	if(!is_attachment(attacking))
+		return
+
+	var/list/attachment_data = get_attachment_data(attacking)
+
+	if(!do_attach(attacking, attacker, attachment_data))
+		return
+
+	finish_handle_attachment(attacking, attachment_data, attacker)
+	var/mob/living/carbon/human/human_attacker = attacker
+	human_attacker.temporarilyRemoveItemFromInventory(attacking)
+
+///Finishes setting up the attachment. This is where the attachment actually attaches. This can be called directly to bypass any checks to directly attach an object.
+/datum/component/attachment_handler/proc/finish_handle_attachment(obj/item/attachment, list/input_attachment_data, mob/attacker)
+	var/list/attachment_data = input_attachment_data
+	if(!input_attachment_data)
+		attachment_data = get_attachment_data(attachment)
+
+	if(slots[attachment_data["slot"]])
+		var/obj/item/current_attachment = slots[attachment_data["slot"]]
+		finish_detach(current_attachment, get_attachment_data(current_attachment), attacker)
+
+	attachment.forceMove(parent)
+	slots[attachment_data["slot"]] = attachment
+
+	on_attach?.Invoke(attachment, attacker)
+	var/datum/callback/attachment_on_attach = attachment_data["on_attach"]
+	attachment_on_attach?.Invoke(parent, attacker)
+
+	update_overlay_icon_state(attachment, attachment_data["overlay_icon_state"])
+
+	update_parent_overlay()
+
+///This is the do_after and user checks for attaching.
+/datum/component/attachment_handler/proc/do_attach(obj/item/attachment, mob/attacher, list/attachment_data)
+	if(!ishuman(attacher))
+		return FALSE
```

Comment:
does this REALLY have to be a human ?

Reply:
I put the check in with habit. Is there any particular reason I should remove it?